File: 22-GenerateParentheses/22-GenerateParentheses.java
// Last updated: 2/3/2026, 9:44:03 PM
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> list =new ArrayList<>();
        backtrack(list,0,0,"",n);
        return list;

    }
    static void backtrack(List<String> res,int open,int close,String curr,int n){
        if(curr.length()==n*2){
            res.add(curr);
            return;
        }
        if(open <n) backtrack(res,open+1,close,curr+"(",n);
        if(open>close) backtrack(res,open,close+1,curr+")",n);
    }
}

File: 22-GenerateParentheses/22-GenerateParentheses.py
# Last updated: 2/3/2026, 9:43:53 PM
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res,sol=[],[]
        def dfs(open_b,close_b):
            if len(sol)==2*n:
                res.append("".join(sol.copy()))
                return
            
            if open_b<n:
                sol.append("(")
                dfs(open_b+1,close_b)
                sol.pop()

            if close_b<open_b:
                sol.append(")")
                dfs(open_b,close_b+1)
                sol.pop()

        dfs(0,0)
        return res